docs/developer/decision-records/2024-10-09-vault-authentication-refactor/README.md

```diff
@@ -0,0 +1,244 @@
+
+### Haschicorp Vault Auth Service Extension
+
+For every authentication method, an implementation of the [Hashicorp Vault Auth interface](#hashicorp-vault-auth-interface) is needed.
+Each `HashicorpVaultAuth` implementation represents a different authentication method and is packaged inside its own service extension.
+In this way, it can easily be added/removed from the runtime and maintained in one place.
+Due to the possible differences in the needed configuration of different authentication methods, each Service Extension will need its own configuration values specific to the authentication method.
+
+### Simple Token Auth
+
+To keep the HashiCorp Vault Extension functional on its own for demo and testing purposes, `SimpleTokenAuth` is implemented.
+`SimpleTokenAuth` is a very rudimentary implementation of the [Hashicorp Vault Auth interface](#hashicorp-vault-auth-interface), that only stores a token and does not offer any refresh functionality.
+It is always added to the registry  and will be used when `fallbackToken` is selected as authentication method, more on that in the [Configuration](#configuration) section.
+
+```java
+public class SimpleTokenAuth implements HashicorpVaultAuth {
+    private String vaultToken;
+
+    SimpleTokenAuth(String vaultToken) {
+        this.vaultToken = vaultToken;
+    }
+
+    @Override
+    public String vaultToken() {
+        return vaultToken;
+    }
+}
+```
+
+### Hashicorp Vault Auth Registry
+
+In line with the registry pattern, `HashicorpVaultAuthRegistry` and `HashicorpVaultAuthRegistryImpl` are created.
+The `HashicorpVaultAuthRegistry` will be used to store one or more implementations of `HashicorpVaultAuth`, each representing a different authentication method.
+More on the usage of the `HashicorpVaultAuthRegistry` for registration in the [HashiCorp Vault Auth Registration](#hashicorp-vault-auth-registration) section.
+The `HashicorpVaultAuthRegistryImpl` is added to the constructor of `HashicorpVaultClient` and can then be used to retrieve the `vaultToken` for the Header creation, more in that in the [HashiCorp Vault Client](#hashicorp-vault-client) section.
+
+```java
+public interface HashicorpVaultAuthRegistry {
+    
+    void register(String method, HashicorpVaultAuth authImplementation);
+    
+    @NotNull
+    HashicorpVaultAuth resolve(String method);
+    
+    boolean hasService(String method);
+}
+```
+
+```java
+public class HashicorpVaultAuthRegistryImpl implements HashicorpVaultAuthRegistry {
+    
+    private final Map<String, HashicorpVaultAuth> services = new HashMap<>();
+
+    public HashicorpVaultAuthRegistryImpl() {
+    }
+
+    @Override
+    public void register(String context, HashicorpVaultAuth service) {
+        services.put(method, service);
+    }
+
+    @Override
+    public @NotNull HashicorpVaultAuth resolve(String method) {
+        return services.get(method);
+    }
```

Comment:
the `@NotNull` annotation is not satisfied by the return statement, because the `get` method can return `null`.
A possibility here would be to throw an `IllegalArgumentException` in case the auth object is null.